### DISCUSSION TITLE
Is there any way to determine whether a component is a client-side component or a server-side component?

### DISCUSSION THREAD

**Discussion by ideaviewes** [vercel/next.js#63745] [App Router]
Is there any way to determine whether a component is a client-side component or a server-side component? Can you console.log it out?

**Comment by icyJoseph** [comment]
AFAIK, at the moment such tooling doesn't exist. I haven't look around community tools.

That being said: 

- All components are first, server components, unless opted-in
- Components that use client side javascript (use-effect, event handlers, state), must be client components 

Also, the component wording here is a bit misleading, it is the JavaScript Module (file if you will) that gets turned into either a Server Module (default), or Client Module (bundled to be sent to the client, with 'use client').

With that in mind, it is possible for a module, call it `A.js`, without a `use client` directive at the top, to be imported by a module that, does have a `use client` directive, say `B.js`, and that'll make `A.js` a client module.

So, this goes beyond the JavaScript expressions, into the bundler territory. I guess it'd be possible for bundlers to make available certain ENV vars to tell you under which situation you are, but that AFAIK is not available yet, and it would require building a module dependency tree first, and then evaluate the JavaScript AST injecting some kind of information about the module. Again, I haven't investigated too much. 

I also think that using this kind of meta analysis, of whether a module will be sent to the client, or not, and based on that, make some code decisions opens up for way more complexity that necessary.

If you want to enforce a module's behavior, you can use these two packages:

- 'server-only', https://www.npmjs.com/package/server-only
- 'client-only', https://www.npmjs.com/package/client-only

Placing this at the top of a module, will "poison"/"taint" the module, so that at development time, you'll see when a module is used incorrectly.
